[PyTorch Mobile] Skip inferring function schema from the C++ function type

dhruvbird opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50457 [PyTorch Mobile] Skip inferring function schema from the C++ function type**

The code to infer function schema from a C++ function relies on templates and code expansion. This uses valuable binary size. We can avoid inferring the schema from the C++ function type (arguments, name, return value) in case that the function implementation is being added to the dispatcher via `m.impl`. In this case, it is assumed that we have a schema registered already. Adding an implementation via `m.def` still triggers schema inferrence.

In addition, we don't do schema schema checks on mobile, so the schema is not needed in the first place.

Differential Revision: [D25853094](https://our.internmc.facebook.com/intern/diff/D25853094/)

dhruvbird committed:
[PyTorch Mobile] Skip inferring function schema from the C++ function type

The code to infer function schema from a C++ function relies on templates and code expansion. This uses valuable binary size. We can avoid inferring the schema from the C++ function type (arguments, name, return value) in case that the function implementation is being added to the dispatcher via `m.impl`. In this case, it is assumed that we have a schema registered already. Adding an implementation via `m.def` still triggers schema inferrence.

In addition, we don't do schema schema checks on mobile, so the schema is not needed in the first place.

Differential Revision: [D25853094](https://our.internmc.facebook.com/intern/diff/D25853094/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch Mobile] Skip inferring function schema from the C++ function type"


The code to infer function schema from a C++ function relies on templates and code expansion. This uses valuable binary size. We can avoid inferring the schema from the C++ function type (arguments, name, return value) in case that the function implementation is being added to the dispatcher via `m.impl`. In this case, it is assumed that we have a schema registered already. Adding an implementation via `m.def` still triggers schema inferrence.

In addition, we don't do schema schema checks on mobile, so the schema is not needed in the first place.

Differential Revision: [D25853094](https://our.internmc.facebook.com/intern/diff/D25853094/)

[ghstack-poisoned]

smessmer reviewed (approved):
does this pass CI? I would have expected it to fail in the schema check when the schema is nullptr.

dhruvbird commented:
> does this pass CI? I would have expected it to fail in the schema check when the schema is nullptr.

I think https://github.com/pytorch/pytorch/pull/49469 stopped making that check for mobile - lmk if it adds up.

smessmer commented:
That's the check at call time which doesn't need schema inference but just checks that the kernel function pointer matches the exact C++ arguments passed at the call site (without going through function schema). I was talking about `checkSchema` in `OperatorEntry.cpp`, but I just checked and it seems nullptr is explicitly handled, so you're fine.

dhruvbird committed:
Update on "[PyTorch Mobile] Skip inferring function schema from the C++ function type"


The code to infer function schema from a C++ function relies on templates and code expansion. This uses valuable binary size. We can avoid inferring the schema from the C++ function type (arguments, name, return value) in case that the function implementation is being added to the dispatcher via `m.impl`. In this case, it is assumed that we have a schema registered already. Adding an implementation via `m.def` still triggers schema inferrence.

In addition, we don't do schema schema checks on mobile, so the schema is not needed in the first place.

Differential Revision: [D25853094](https://our.internmc.facebook.com/intern/diff/D25853094/)

[ghstack-poisoned]

dhruvbird committed:
Update on "[PyTorch Mobile] Skip inferring function schema from the C++ function type"


The code to infer function schema from a C++ function relies on templates and code expansion. This uses valuable binary size. We can avoid inferring the schema from the C++ function type (arguments, name, return value) in case that the function implementation is being added to the dispatcher via `m.impl`. In this case, it is assumed that we have a schema registered already. Adding an implementation via `m.def` still triggers schema inferrence.

In addition, we don't do schema schema checks on mobile, so the schema is not needed in the first place.

Differential Revision: [D25853094](https://our.internmc.facebook.com/intern/diff/D25853094/)

[ghstack-poisoned]
